WordPress/src/main/java/org/wordpress/android/models/discover/RecommendedBlog.kt

```diff
@@ -0,0 +1,11 @@
+package org.wordpress.android.models.discover
+
+data class RecommendedBlog(
+    val name: String,
+    val url: String,
+    val blogId: Long,
+    val description: String,
+    val iconUrl: String?,
+    val feedId: Long?,
+    val isFollowed: Boolean
+)
```

Comment:
I've decided to create a new model that will combine information from two sources:
1. Discovery endpoint - blog details
2. Blog Table - whether a blog is followed or not

I don't want to store `isFollowed` information alongside recommended_blogs card because this would create another source of truth instead of having one (Blog table). I believe this is done like that with the Posts and this complicates the logic when following the Site. Take a look [here](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/reader/actions/ReaderBlogActions.java#L197). If I would follow this pattern we would have to add third line to store the same information. 

So with the current implementation the data flow would look like this:
1. Fetch `recommended_blogs` 
2. Store it locally - currently this is stored as plain json, so there's no info about the follow state
3. Get Discovery Cards -> here we retrieve the recommended blogs and map it to the new model where we additionally check the information whether a site is followed or not in the Blog table

thanks to this when we follow a Site we don't have to store it in a separate places.

Reply:
> I believe this is done like that with the Posts and this complicates the logic when following the Site

Can you clarify this a bit? And what difference in pattern do you see between following a post site (using more menu on the top right of the discover card) vs following a recommended site.

Reply:
Sure! 
> And what difference in pattern do you see between following a post site (using more menu on the top right of the discover card) vs following a recommended site.

I think there should be no difference. It shouldn't matter where the `follow` button is clicked. That's why I did try to decouple it from `post` a bit by changing the params and function names in the code. 

The comment above was about the internal storing logic. Currently, whether a blog is followed or not is stored in two tables: `blog` & `post`. When you `follow/unfollow` a blog you need to update both tables. I would guess that [doing this for each post](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/datasets/ReaderPostTable.java#L756) stored locally comes with a small performance penalty. 

The approach I took here was to treat `Blog` table as the single source of truth for information about `follow` state. So instead of keeping `isFollow` locally in another table where recommended blogs are stored I combining this two information later on. 
This could also be done with Posts. Each Post object instead of keeping `isFollow` info could just keep a foreign key to the blog object. When loading posts we could build the full object with `isFollow` info by retrieving it from the `Blog` table. Thanks to this we would only have to store `isFollow` in one table.

Reply:
Thanks for the explanation :). I still have few questions regarding the original comment:

> I don't want to store isFollowed information alongside recommended_blogs card

> we would have to add third line to store the same information.

Haven’t we already created two sources for blogs data by saving recommended blogs json data into `tbl_discover_cards`?

I see that `score` field is returned in the json that cannot be inserted into `tbl_blog_info`. Similarly there are some missing fields from `tbl_blog_info` not being returned from the API for the `recommended_blog` card. 

Considering we’ve just few boolean `blog` fields that are missing from the `recommended_blog` response, if we receive these missing `blog` fields and ignore `score` field, how do you feel about keeping single data source for blogs by saving `recommended_blog` to `tbl_blog_info`? That ways we can reuse existing logic to update follow state and need not add a "third line to store the same information" or create new models.

I'm just trying to understand possibilities, feel free to put your arguments.

> Each Post object instead of keeping isFollow info could just keep a foreign key to the blog object. 

Totally agree. Let's keep it out of scope of this task at the moment :).

Reply:
> Haven’t we already created two sources for blogs data by saving recommended blogs json data into tbl_discover_cards?

Kinda yes, this may be the lack of knowledge of the project on my side but I thought only followed blogs are stored locally and I treated the "recommended blog" as a different thing. Additionally, we still want to store `recommended_card` separately as I believe the Discovery view is driven by the local storage.

> I see that score field is returned in the json that cannot be inserted into tbl_blog_info.

Yeah, but I'm not parsing that field as it's not needed, so I don't think that would be the problem for now. 

> Considering we’ve just few boolean blog fields that are missing from the recommended_blog response, if we receive these missing blog fields and ignore score field, how do you feel about keeping single data source for blogs by saving recommended_blog to tbl_blog_info?

Yeah I think we could do that as well. We would still need to store some information about the `recommended_blogs` card though.

> That ways we can reuse existing logic to update follow state and need not add a "third line to store the same information" or create new models.
We can reuse the current logic now? At least that's what I did that and it seems to work, or am I missing something?


So to sum up, I would say that we can use the  `tbl_blog_info` as long as we are not treating the recommended blog as a different thing and we are fine with storing a lot of blogs over time. 
If I was to make that decision(with my current knowledge about the project) I would probably stay with the current implementation as it keeps `recommended_blogs` in a separate place but it still combines `isFollow` information from the `tbl_blog_info` so no need for extra logic. 

One thing that I saw right now when writing this comment and going through the code is that there's `tbl_recommended_blogs` table but this one seems to be used specifically for one view as it drops all records when updating new ones, so that would

Reply:
> We can reuse the current logic now? At least that's what I did that and it seems to work, or am I missing something?

My comment was meant for this part of your answer:
> I believe this is done like that with the Posts and this complicates the logic when following the Site. 

If it shouldn't matter where the follow button is clicked, logic should not get complicated when following the Site.

Let’s continue with the discussion once you’ve created a PR in the main repo :).